sketch/src/sketch_rnn/model.ts

```diff
@@ -0,0 +1,359 @@
+   * @param softmaxTemperature (Optional) for Pi and Pen discrete states
+   * (default is temperature * 0.5 + 0.5, which is a nice heuristic.)
+   *
+   * @returns [pi, mu1, mu2, sigma1, sigma2, corr, pen]
+   */
+  getPDF(state: Float32Array[],
+    temperature=0.65,
+    softmaxTemperature?: number) {
+    const temp = temperature;
+    let discreteTemp: number = 0.5 + temp * 0.5; // good heuristic.
+    if (softmaxTemperature) {
+      discreteTemp = softmaxTemperature;
+    }
+    const out = tf.tidy(() => {
+      const numUnits = this.numUnits;
+      const h = tf.tensor2d(state[1], [1, numUnits]);
+      const NOUT = this.NMIXTURE;
+
+      const sqrttemp = tf.scalar(Math.sqrt(temp));
+      const softtemp = tf.scalar(discreteTemp);
+
+      const z = tf.add(tf.matMul(h, this.outputKernel), this.outputBias)
+        .squeeze();
+
+      const [rawPen, rst] = tf.split(z, [3, NOUT*6]);
+      const [rawPi, mu1, mu2, rawSigma1, rawSigma2, rawCorr] = tf.split(rst, 6);
+      const pen = tf.softmax(rawPen.div(softtemp));
+      const pi = tf.softmax(rawPi.div(softtemp));
+      const sigma1 = tf.exp(rawSigma1).mul(sqrttemp);
+      const sigma2 = tf.exp(rawSigma2).mul(sqrttemp);
+      const corr = tf.tanh(rawCorr);
+      return [pi, mu1, mu2, sigma1, sigma2, corr, pen];
+    });
+    const result = [];
+    for (let i = 0; i < out.length; i++) {
+      result.push(out[i].dataSync());
+      out[i].dispose();
+    }
+    return result;
+  }
+
+  /**
+   * Returns the zero/initial state of the model
+   *
+   * @returns zero state of the lstm: [c, h], where c and h are zero vectors.
+   */
+  zeroState() {
+    const result: Float32Array[] = [
+      new Float32Array(this.numUnits),
+      new Float32Array(this.numUnits),
+    ];
+    return result;
+  }
+
+  /**
+   * Returns a new copy of the rnn state
+   *
+   * @param rnnState original [c, h] states of the lstm
+   *
+   * @returns copy of state of the lstm: [c, h]
+   */
+  copyState(rnnState: Float32Array[]) {
+    const c = new Float32Array(rnnState[0]);
+    const h = new Float32Array(rnnState[1]);
+    return [c, h];
+  }
+
+  /**
+   * Returns the zero input state of the model
+   *
+   * @returns [0, 0, 1, 0, 0].
+   */
+  zeroInput() {
+    return [0, 0, 1, 0, 0];
+  }
+
+  /**
+   * Samples the next point of the sketch given pdf parameters
+   *
+   * @param pdf result from getPDF() call
+   *
+   * @returns [dx, dy, penDown, penUp, penEnd]
+   */
+  sample(pdf: number[][]) {
+    // pdf is [pi, mu1, mu2, sigma1, sigma2, corr, pen]
+    // returns [x, y, eos]
+    const idx = support.sampleSoftmax(pdf[0]);
+    const mu1 = pdf[1][idx];
+    const mu2 = pdf[2][idx];
+    const sigma1 = pdf[3][idx];
+    const sigma2 = pdf[4][idx];
+    const corr = pdf[5][idx];
+    const penIdx = support.sampleSoftmax(pdf[6]);
+    const penstate = [0, 0, 0];
+    penstate[penIdx] = 1;
+    const delta = support.birandn(mu1, mu2, sigma1, sigma2, corr);
+    return [
```

Comment:
Again, I think it's a good idea to return an object (with a defined interface).

Reply:
good suggestion, agree with object/interface for this.